xarray/core/merge.py

```diff
@@ -360,7 +360,8 @@ def merge_data_and_coords(data, coords, compat='broadcast_equals',
     """Used in Dataset.__init__."""
     objs = [data, coords]
     explicit_coords = coords.keys()
-    return merge_core(objs, compat, join, explicit_coords=explicit_coords)
+    return merge_core(objs, compat, join, explicit_coords=explicit_coords,
+                      indexes=coords)
```

Comment:
My concern was passing non-indices here. I'm still not 100% why `indexes` is tolerant to arguments that are index coordinates but it seems like it works.